File: 03-系统编程/04-文件系统/完整性校验/python/verity.py
"""fs-verity Merkle 树 + dm-integrity 布局的可计算模型。

来源（全部实际读过，见 README）：
* ``Documentation/filesystems/fsverity.html``   —— 分块、salt 填充、descriptor
* ``include/uapi/linux/fsverity.h``             —— 常量与 ioctl 结构
* ``Documentation/admin-guide/device-mapper/dm-integrity.html`` —— 盘上布局
"""

import hashlib

# ---- include/uapi/linux/fsverity.h ----------------------------------------
FS_VERITY_HASH_ALG_SHA256 = 1
FS_VERITY_HASH_ALG_SHA512 = 2

# 最大 salt 长度：文档写 "Currently the maximum salt size is 32 bytes"，
# 与 descriptor 里的 __u8 salt[32] 一致（常量名在内核内部头里，未取到原文）
FS_VERITY_MAX_SALT_SIZE = 32

# fsverity.h 的 FS_VERITY_METADATA_TYPE_*
FS_VERITY_METADATA_TYPE_MERKLE_TREE = 1
FS_VERITY_METADATA_TYPE_DESCRIPTOR = 2
FS_VERITY_METADATA_TYPE_SIGNATURE = 3

# ---- Documentation/filesystems/fsverity.html -------------------------------
FSVERITY_DESCRIPTOR_SIZE = 256
FSVERITY_ROOT_HASH_SIZE = 64        # struct fsverity_descriptor.root_hash[64]
FSVERITY_SALT_FIELD = 32            # struct fsverity_descriptor.salt[32]

# 各算法"压缩函数输入块"大小，salt 要零填充到它的整数倍
HASH_ALG_BLOCKSIZE = {
    FS_VERITY_HASH_ALG_SHA256: 64,
    FS_VERITY_HASH_ALG_SHA512: 128,
}
HASH_ALG_DIGESTSIZE = {
    FS_VERITY_HASH_ALG_SHA256: 32,
    FS_VERITY_HASH_ALG_SHA512: 64,
}
HASH_ALG_NAME = {
    FS_VERITY_HASH_ALG_SHA256: "sha256",
    FS_VERITY_HASH_ALG_SHA512: "sha512",
}

# ---- fsverity descriptor 字段偏移（小端）----------------------------------
# version(1) hash_algorithm(1) log_blocksize(1) salt_size(1)
# __reserved_0x04(4) data_size(8) root_hash(64) salt(32) __reserved[144]
DESC_OFF_VERSION = 0
DESC_OFF_HASH_ALG = 1
DESC_OFF_LOG_BLOCKSIZE = 2
DESC_OFF_SALT_SIZE = 3
DESC_OFF_RESERVED_0x04 = 4
DESC_OFF_DATA_SIZE = 8
DESC_OFF_ROOT_HASH = 16
DESC_OFF_SALT = 16 + 64
DESC_OFF_RESERVED = 16 + 64 + 32


def pad_salt(salt: bytes, alg: int) -> bytes:
    """盐零填充到压缩函数输入块大小的整数倍。

    文档原话："the salt is zero-padded to the closest multiple of the input
    size of the hash algorithm's compression function, e.g. 64 bytes for
    SHA-256 or 128 bytes for SHA-512"。
    这样做的目的是让每个 hash 的输入长度一致（便于硬件加速 / 预计算状态）。
    """
    bs = HASH_ALG_BLOCKSIZE[alg]
    if not salt:
        return b""
    n = (len(salt) + bs - 1) // bs * bs
    return salt + b"\x00" * (n - len(salt))


def merkle_level_sizes(data_size: int, block_size: int, digest_size: int):
    """返回每一层"有多少个 block"，第 0 层是数据块。

    文档：把文件内容按 block_size 切块，末尾零填充；每块算 hash 得到第一层；
    第一层的 hash 再按 block_size 分组、零填充、再 hash，直到只剩一块。
    """
    if data_size == 0:
        return []
    nblocks = (data_size + block_size - 1) // block_size
    levels = [nblocks]
    hashes = nblocks * digest_size
    while True:
        if nblocks == 1:
            break
        nblocks = (hashes + block_size - 1) // block_size
        levels.append(nblocks)
        hashes = nblocks * digest_size
    return levels


def arity(block_size: int, digest_size: int) -> int:
    """每个 block 能装多少个下层 hash。"""
    return block_size // digest_size


def merkle_tree_overhead(data_size: int, block_size: int, digest_size: int) -> int:
    """Merkle 树本身的字节数（不含 descriptor）。

    大文件收敛到约 1/(arity-1) 的原文大小：SHA-256 + 4K 块即 1/127。
    """
    levels = merkle_level_sizes(data_size, block_size, digest_size)
    if not levels:
        return 0
    return sum(n * block_size for n in levels[1:])


def _hash(data: bytes, alg: int) -> bytes:
    return hashlib.new(HASH_ALG_NAME[alg], data).digest()


def build_merkle_tree(data: bytes, block_size: int, alg: int, salt: bytes = b""):
    """返回 (root_hash, levels)，levels[i] 是第 i 层的 block 列表。

    第 0 层是原文分块（末尾零填充）；第 i+1 层是第 i 层各块 hash 拼接后
    再分块再 hash。空文件的 root 是全 0。
    """
    ds = HASH_ALG_DIGESTSIZE[alg]
    psalt = pad_salt(salt, alg)
    if not data:
        return b"\x00" * ds, []

    def blk(i):
        return data[i * block_size:(i + 1) * block_size]

    n0 = (len(data) + block_size - 1) // block_size
    level0 = [blk(i).ljust(block_size, b"\x00") for i in range(n0)]
    levels = [level0]
    cur = level0
    while True:
        digests = b"".join(_hash(psalt + b, alg) for b in cur)
        if len(cur) == 1:
            # 只剩一块：这一层的单个 hash 就是 root
            # （若文件本身只占一块，root 就是那块数据的 hash）
            return digests[:ds], levels
        packed = [digests[i * block_size:(i + 1) * block_size].ljust(block_size, b"\x00")
                  for i in range((len(digests) + block_size - 1) // block_size)]
        levels.append(packed)
        cur = packed


def build_descriptor(data_size: int, log_blocksize: int, alg: int,
                     root_hash: bytes, salt: bytes = b"") -> bytes:
    """struct fsverity_descriptor 的 256 字节小端序列化。"""
    d = bytearray(FSVERITY_DESCRIPTOR_SIZE)
    d[DESC_OFF_VERSION] = 1
    d[DESC_OFF_HASH_ALG] = alg
    d[DESC_OFF_LOG_BLOCKSIZE] = log_blocksize
    d[DESC_OFF_SALT_SIZE] = len(salt)
    d[DESC_OFF_DATA_SIZE:DESC_OFF_DATA_SIZE + 8] = data_size.to_bytes(8, "little")
    d[DESC_OFF_ROOT_HASH:DESC_OFF_ROOT_HASH + len(root_hash)] = root_hash
    d[DESC_OFF_SALT:DESC_OFF_SALT + len(salt)] = salt
    return bytes(d)


def file_digest(descriptor: bytes, alg: int) -> bytes:
    """fs-verity 的"文件摘要"= hash(descriptor)。

    关键点：文件摘要**不是** Merkle root 本身。root 有歧义——无法区分
    "一个大文件"和"一个内容恰好等于前者顶层 hash 块的小文件"。
    """
    return _hash(descriptor, alg)


def verify_block(data: bytes, index: int, root: bytes, block_size: int,
                 alg: int, salt: bytes = b"", levels=None) -> bool:
    """给定块的兄弟路径，验证其 hash 链能回到 root。

    这里直接用完整 levels 做（真实内核只保存需要的那部分），
    重点是把"路径长度 == 层数-1"这条性质钉住。
    """
    ds = HASH_ALG_DIGESTSIZE[alg]
    psalt = pad_salt(salt, alg)
    if levels is None:
        _, levels = build_merkle_tree(data, block_size, alg, salt)
    a = arity(block_size, ds)
    h = _hash(psalt + levels[0][index], alg)
    idx = index
    for lv in range(1, len(levels)):
        sib_off = (idx // a) * a
        packed = b"".join(_hash(psalt + levels[lv - 1][j], alg)
                          for j in range(sib_off, min(sib_off + a, len(levels[lv - 1]))))
        h = _hash(psalt + packed.ljust(block_size, b"\x00"), alg)
        idx //= a
    return h == root


# --------------------------------------------------------------------------
# dm-integrity 盘上布局
# --------------------------------------------------------------------------
DM_INTEGRITY_DEFAULT_INTERLEAVE_SECTORS = 32768
DM_INTEGRITY_DEFAULT_BUFFER_SECTORS = 128
DM_INTEGRITY_DEFAULT_JOURNAL_WATERMARK = 50
DM_INTEGRITY_DEFAULT_COMMIT_TIME_MS = 10000
DM_INTEGRITY_BLOCK_SIZES = (512, 1024, 2048, 4096)
DM_INTEGRITY_DEFAULT_BLOCK_SIZE = 512
DM_INTEGRITY_SUPERBLOCK_SECTORS = 8          # 4 KiB / 512
DM_INTEGRITY_JOURNAL_ENTRY_DATA = 504        # 每扇区存 504 字节，末 8 字节在 journal
SECTOR_SIZE = 512


def rounddown_pow2(x: int) -> int:
    """interleave_sectors / buffer_sectors 都是"向下取到 2 的幂"。"""
    if x <= 0:
        return 0
    return 1 << (x.bit_length() - 1)


def integrity_geometry(interleave_sectors: int, block_size: int, tag_size: int,
                       buffer_sectors: int = DM_INTEGRITY_DEFAULT_BUFFER_SECTORS):
    """算一个 interleave run 里"标签区 + 数据区"的扇区布局。

    文档示例：interleave_sectors=32768、block_size=512、crc32c tag=4 字节
    → 一整个数据区需要 32768*4 = 128 KiB 标签 = 256 个扇区；
    buffer_sectors=128 → 每个元数据区 2 个 buffer，即每 16 MiB 数据 2 个 buffer。
    """
    il = rounddown_pow2(interleave_sectors)
    # 标签区至少 4 KiB，向上取整到扇区
    tag_bytes = il * tag_size
    tag_sectors = (tag_bytes + SECTOR_SIZE - 1) // SECTOR_SIZE
    # 标签区本身要占掉 interleave 区域里的扇区，且它自己也要 4KiB 对齐
    tag_sectors = max(tag_sectors, SECTOR_SIZE // SECTOR_SIZE * 8)   # >= 4 KiB
    data_sectors = il - tag_sectors
    # 数据扇区数必须是 2 的幂（log2 存在 superblock 里）
    data_sectors = rounddown_pow2(data_sectors)
    bufs = (tag_sectors + buffer_sectors - 1) // buffer_sectors
    return {
        "interleave_sectors": il,
        "tag_bytes": tag_bytes,
        "tag_sectors": tag_sectors,
        "data_sectors": data_sectors,
        "buffers_per_run": bufs,
        "data_bytes": data_sectors * SECTOR_SIZE,
        "overhead_ratio": tag_sectors / (tag_sectors + data_sectors),
    }

File: 03-系统编程/04-文件系统/完整性校验/go/verity.go
// fs-verity Merkle 树与 dm-integrity 布局的 Go 侧镜像。
//
// 来源同 Python 侧：Documentation/filesystems/fsverity.html、
// include/uapi/linux/fsverity.h、
// Documentation/admin-guide/device-mapper/dm-integrity.html
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
)

const (
	fsVerityHashAlgSha256 = 1
	fsVerityHashAlgSha512 = 2

	fsVerityMaxSaltSize = 32

	fsverityDescriptorSize = 256
	fsverityRootHashSize   = 64

	descOffVersion      = 0
	descOffHashAlg      = 1
	descOffLogBlocksize = 2
	descOffSaltSize     = 3
	descOffDataSize     = 8
	descOffRootHash     = 16
	descOffSalt         = 80

	blockSize = 4096
	sha256DS  = 32

	// dm-integrity 默认值
	dmDefaultInterleaveSectors = 32768
	dmDefaultBufferSectors     = 128
	dmDefaultBlockSize         = 512
	sectorSize                 = 512
)

// hashAlgBlocksize 是各算法压缩函数的输入块大小（salt 要填充到它的整数倍）。
var hashAlgBlocksize = map[int]int{
	fsVerityHashAlgSha256: 64,
	fsVerityHashAlgSha512: 128,
}

var hashAlgDigestsize = map[int]int{
	fsVerityHashAlgSha256: 32,
	fsVerityHashAlgSha512: 64,
}

// PadSalt 把 salt 零填充到压缩函数输入块的整数倍。
func PadSalt(salt []byte, alg int) []byte {
	if len(salt) == 0 {
		return nil
	}
	bs := hashAlgBlocksize[alg]
	n := (len(salt) + bs - 1) / bs * bs
	out := make([]byte, n)
	copy(out, salt)
	return out
}

// Arity 是每个块能装的下层 hash 个数。
func Arity(blockSize, digestSize int) int { return blockSize / digestSize }

// MerkleLevelSizes 返回每层有多少块，第 0 层是数据块。
func MerkleLevelSizes(dataSize, blockSize, digestSize int) []int {
	if dataSize == 0 {
		return nil
	}
	nblocks := (dataSize + blockSize - 1) / blockSize
	levels := []int{nblocks}
	hashes := nblocks * digestSize
	for nblocks != 1 {
		nblocks = (hashes + blockSize - 1) / blockSize
		levels = append(levels, nblocks)
		hashes = nblocks * digestSize
	}
	return levels
}

// MerkleTreeOverhead 是 Merkle 树本身的字节数（不含 descriptor）。
func MerkleTreeOverhead(dataSize, blockSize, digestSize int) int {
	lv := MerkleLevelSizes(dataSize, blockSize, digestSize)
	if len(lv) < 2 {
		return 0
	}
	sum := 0
	for _, n := range lv[1:] {
		sum += n * blockSize
	}
	return sum
}

func sha256Of(b []byte) []byte {
	s := sha256.Sum256(b)
	return s[:]
}

// BuildDescriptor 序列化 struct fsverity_descriptor（小端）。
func BuildDescriptor(dataSize, logBlocksize, alg int, rootHash, salt []byte) []byte {
	d := make([]byte, fsverityDescriptorSize)
	d[descOffVersion] = 1
	d[descOffHashAlg] = byte(alg)
	d[descOffLogBlocksize] = byte(logBlocksize)
	d[descOffSaltSize] = byte(len(salt))
	for i := 0; i < 8; i++ {
		d[descOffDataSize+i] = byte(uint64(dataSize) >> uint(8*i))
	}
	copy(d[descOffRootHash:descOffRootHash+len(rootHash)], rootHash)
	copy(d[descOffSalt:descOffSalt+len(salt)], salt)
	return d
}

// FileDigest 是 hash(descriptor)，不是 Merkle root。
func FileDigest(descriptor []byte) []byte { return sha256Of(descriptor) }

// BuildMerkleRoot 建树并返回 root 与层数。
func BuildMerkleRoot(data []byte, bs, alg int) ([]byte, int) {
	ds := hashAlgDigestsize[alg]
	if len(data) == 0 {
		return make([]byte, ds), 0
	}
	n0 := (len(data) + bs - 1) / bs
	cur := make([][]byte, 0, n0)
	for i := 0; i < n0; i++ {
		b := make([]byte, bs)
		copy(b, data[i*bs:min(i*bs+bs, len(data))])
		cur = append(cur, b)
	}
	levels := 1
	for {
		var digests []byte
		for _, b := range cur {
			digests = append(digests, sha256Of(b)...)
		}
		if len(cur) == 1 {
			return digests[:ds], levels
		}
		packed := make([][]byte, 0)
		for i := 0; i < len(digests); i += bs {
			b := make([]byte, bs)
			copy(b, digests[i:min(i+bs, len(digests))])
			packed = append(packed, b)
		}
		cur = packed
		levels++
	}
}

// RounddownPow2 用于 interleave_sectors / buffer_sectors 的向下取整。
func RounddownPow2(x int) int {
	if x <= 0 {
		return 0
	}
	return 1 << (bits(x) - 1)
}

func bits(x int) int {
	n := 0
	for x > 0 {
		x >>= 1
		n++
	}
	return n
}

// Geometry 是一个 interleave run 的扇区布局。
type Geometry struct {
	TagBytes      int
	TagSectors    int
	DataSectors   int
	BuffersPerRun int
	Overhead      float64
}

// IntegrityGeometry 计算 dm-integrity 的标签区/数据区布局。
func IntegrityGeometry(interleaveSectors, blockSize, tagSize, bufferSectors int) Geometry {
	il := RounddownPow2(interleaveSectors)
	tagBytes := il * tagSize
	tagSectors := (tagBytes + sectorSize - 1) / sectorSize
	if tagSectors < 8 { // 标签区至少 4 KiB
		tagSectors = 8
	}
	dataSectors := RounddownPow2(il - tagSectors)
	bufs := (tagSectors + bufferSectors - 1) / bufferSectors
	return Geometry{
		TagBytes:      tagBytes,
		TagSectors:    tagSectors,
		DataSectors:   dataSectors,
		BuffersPerRun: bufs,
		Overhead:      float64(tagSectors) / float64(tagSectors+dataSectors),
	}
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	fmt.Println("== salt 零填充（填充到压缩函数输入块的整数倍） ==")
	for _, n := range []int{0, 13, 32, 64, 65} {
		s := make([]byte, n)
		fmt.Printf("  %-3d 字节(SHA-256) → %-4d 字节\n", n, len(PadSalt(s, fsVerityHashAlgSha256)))
	}
	fmt.Printf("  13  字节(SHA-512) → %-4d 字节\n",
		len(PadSalt(make([]byte, 13), fsVerityHashAlgSha512)))

	fmt.Println("== 扇出与层级 ==")
	fmt.Printf("  SHA-256 + 4K 块 → arity %d\n", Arity(blockSize, sha256DS))
	fmt.Printf("  SHA-512 + 4K 块 → arity %d\n", Arity(blockSize, 64))
	for _, n := range []int{1, 2, 128, 129, 16384} {
		fmt.Printf("  %-6d 块 → %v\n", n,
			MerkleLevelSizes(n*blockSize, blockSize, sha256DS))
	}

	fmt.Println("== 1/127 收敛 ==")
	for _, n := range []int{2, 128, 16384, 128 * 128 * 128} {
		size := n * blockSize
		oh := MerkleTreeOverhead(size, blockSize, sha256DS)
		fmt.Printf("  %-8d 块 → 树 %-9d 字节，占比 %.4f\n", n, oh, float64(oh)/float64(size))
	}

	fmt.Println("== Merkle root ==")
	cases := []struct {
		name string
		data []byte
	}{
		{"空文件", nil},
		{"1 块", []byte("hello fs-verity")},
		{"2 块", make([]byte, blockSize+10)},
	}
	for _, c := range cases {
		root, lv := BuildMerkleRoot(c.data, blockSize, fsVerityHashAlgSha256)
		fmt.Printf("  %-6s → root=%s... 层数=%d\n", c.name, hex.EncodeToString(root)[:24], lv)
	}
	two := make([]byte, blockSize+10)
	root, _ := BuildMerkleRoot(two, blockSize, fsVerityHashAlgSha256)
	desc := BuildDescriptor(len(two), 12, fsVerityHashAlgSha256, root, nil)
	fmt.Printf("  文件摘要 = sha256(descriptor) = %s...\n",
		hex.EncodeToString(FileDigest(desc))[:24])

	fmt.Println("== dm-integrity 几何 ==")
	for _, cfg := range [][3]int{{512, 4}, {512, 16}, {4096, 32}} {
		g := IntegrityGeometry(dmDefaultInterleaveSectors, cfg[0], cfg[1], dmDefaultBufferSectors)
		fmt.Printf("  block=%-5d tag=%-3d → 标签 %-5d 扇区, 数据 %-6d 扇区,"
			" buffer/区 %-3d 开销 %.4f\n",
			cfg[0], cfg[1], g.TagSectors, g.DataSectors, g.BuffersPerRun, g.Overhead)
	}
}
